ElectricityBillApi.Tests/ElectricProviderPickerServiceTests.cs

```diff
@@ -0,0 +1,166 @@
+﻿using ElectricityBillApi.Models;
+using ElectricityBillApi.Services;
+using NSubstitute;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace ElectricityBillApi.Tests
+{
+    public class ElectricProviderPickerServiceTests
+    {
+        [Fact]
+        public void FindCheapestProviderAndPrice_returnsNearCheapestProvider()
+        {
+            // arrange
+            var providers = GetTestData().ToList();
+            var iProviders = Substitute.For<IElectricProviders>();
+            iProviders.GetProviders.ReturnsForAnyArgs(providers);
+            var sut = new ElectricProviderPickerService(iProviders);
+            var ownAddress = new Address()
+            {
+                Location = new Location()
+                {
+                    X=0,
+                    Y=0,
+                    Z=0
+                }
+            };
+
+            // act 
+            var (cheapestProvider, calculatedPrice) = sut.FindCheapestProviderAndPrice(ownAddress);
+
+            // assert
+            Assert.Equal(providers.First(), cheapestProvider);
+        }
+
+        [Fact]
+        public void FindCheapestProviderAndPrice_returnsFarCheapestProvider()
+        {
+            // arrange
+            var providers = GetTestData().ToList();
+            var iProviders = Substitute.For<IElectricProviders>();
+            iProviders.GetProviders.ReturnsForAnyArgs(providers);
```

Comment:
This is repeated multiple times, can be moved to a ctor for a shared setup. this will not cause a shared state, jut a shared logic for initialization.

Reply:
That's true, I use fields now.